File: src/servers/worker/worker_server.py
"""
Python gRPC Worker Server — Process I
Receives DelegateQuery from parent (Process A), returns empty (owns no counties).
"""

import sys
import os
import json
import time
import logging
import threading
import grpc
from concurrent import futures
from datetime import datetime

# ---------------------------------------------------------------------------
# Proto imports — generated stubs must exist before running this script.
# Run generate_proto.sh first.
# ---------------------------------------------------------------------------
PROTO_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "..", "proto", "generated", "python")
sys.path.insert(0, PROTO_DIR)

import parking_violation_query_pb2       as pb2
import parking_violation_query_pb2_grpc  as pb2_grpc

# ---------------------------------------------------------------------------
# Logging
# ---------------------------------------------------------------------------
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s.%(msecs)03d [%(name)s] %(message)s",
    datefmt="%H:%M:%S",
)
logger = logging.getLogger("WorkerI")

os.makedirs("logs", exist_ok=True)
_log_fh = None  # opened after config is loaded


def _log(event: str, req_id: str = "-", pending: int = 0,
         chunk: int = -1, records: int = -1, extra: str = "") -> None:
    ts = datetime.now().strftime("%H:%M:%S.") + f"{datetime.now().microsecond // 1000:03d}"
    line = (f"{ts} [I] {event:<22} req={req_id} pending={pending}"
            f" chunk={chunk} records={records}")
    if extra:
        line += f" {extra}"
    logger.info(line)
    if _log_fh:
        _log_fh.write(line + "\n")
        _log_fh.flush()


# ---------------------------------------------------------------------------
# Config loader
# ---------------------------------------------------------------------------
def load_config(path: str) -> dict:
    with open(path) as f:
        return json.load(f)


# ---------------------------------------------------------------------------
# Service implementation
# ---------------------------------------------------------------------------
class WorkerServiceImpl(pb2_grpc.ParkingViolationServiceServicer):

    def __init__(self, config: dict):
        self.config = config
        self.process_id = config["process_id"]
        self.owned_counties: list[str] = config["data_partitioning"]["owned_counties"]
        self._pending = 0
        self._pending_lock = threading.Lock()

        logger.info(f"[Worker {self.process_id}] starting (team={config['team']})")
        logger.info(f"[Worker {self.process_id}] listening on "
                    f"{config['listen_host']}:{config['listen_port']}")
        logger.info(f"[Worker {self.process_id}] owned counties: "
                    f"{self.owned_counties if self.owned_counties else '(none)'}")
        _log("INIT", extra="process started")

    # ------------------------------------------------------------------
    def DelegateQuery(self, request, context):
        with self._pending_lock:
            self._pending += 1
            pending = self._pending

        req_id = request.request_id
        delegator = request.delegating_process
        logger.info(f"\n[Worker {self.process_id}] delegation {req_id} from {delegator}")
        _log("RECV_DELEGATION", req_id, pending, extra=delegator)

        # ---- deserialize original QueryRequest ----
        query = pb2.QueryRequest()
        if not query.ParseFromString(request.original_query):
            with self._pending_lock:
                self._pending -= 1
            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
            context.set_details("Failed to parse original_query")
            return

        # ---- determine counties to serve ----
        counties = self._select_counties(query)
        if not counties:
            logger.info(f"[Worker {self.process_id}] no matching counties — returning empty")
            _log("DONE", req_id, pending, chunk=0, records=0)
            with self._pending_lock:
                self._pending -= 1
            return  # yields nothing — empty stream

        # ---- Process I owns no counties in this topology, but the logic
        #      below would work for any worker that actually owns data. ----
        chunk_size = self.config["chunk_config"]["default_chunk_size"]
        records = self._load_data(query, counties)
        logger.info(f"[Worker {self.process_id}] loaded {len(records)} records")
        _log("LOADED", req_id, pending, records=len(records))

        chunk_num = 0
        for i in range(0, len(records), chunk_size):
            if context.is_active() is False:
                _log("CANCELLED", req_id, pending, chunk=chunk_num)
                with self._pending_lock:
                    self._pending -= 1
                return

            batch = records[i:i + chunk_size]
            resp = pb2.DelegationResponse()
            resp.request_id = req_id
            resp.chunk_number = chunk_num
            resp.is_final = False
            resp.responding_process = self.process_id
            for rec in batch:
                resp.records.append(rec)

            yield resp
            _log("CHUNK_SENT", req_id, pending, chunk=chunk_num, records=len(batch))
            logger.info(f"[Worker {self.process_id}] sent chunk {chunk_num} ({len(batch)} records)")
            chunk_num += 1

        _log("DONE", req_id, pending, chunk=chunk_num, records=len(records))
        with self._pending_lock:
            self._pending -= 1

    # ------------------------------------------------------------------
    def HealthCheck(self, request, context):
        with self._pending_lock:
            p = self._pending
        resp = pb2.HealthResponse()
        resp.responding_process = self.process_id
        resp.is_healthy = True
        resp.pending_requests = p
        resp.active_workers = 1
        return resp

    # ------------------------------------------------------------------
    def QueryViolations(self, request, context):
        context.set_code(grpc.StatusCode.UNIMPLEMENTED)
        context.set_details("Workers do not accept direct queries")
        return iter([])

    # ------------------------------------------------------------------
    def CancelQuery(self, request, context):
        resp = pb2.CancelResponse()
        resp.request_id = request.request_id
        resp.cancelled = True
        resp.message = "cancel acknowledged"
        return resp

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------
    def _select_counties(self, query: pb2.QueryRequest) -> list[str]:
        if not self.owned_counties:
            return []
        qc = query.violation_county
        if not qc:
            return list(self.owned_counties)
        if qc in self.owned_counties:
            return [qc]
        return []

    def _load_data(self, query: pb2.QueryRequest,
                   counties: list[str]) -> list[pb2.ViolationRecord]:
        """
        Load matching records from the CSV.
        Returns a list of ViolationRecord proto messages.
        This is a simple sequential scan — suitable for a single-host dev setup.
        """
        data_path = self.config.get("data_path", "")
        if not data_path or not os.path.exists(data_path):
            logger.warning(f"[Worker {self.process_id}] data_path not found: {data_path}")
            return []

        county_set = set(counties)
        vc_filter  = str(query.violation_code) if query.violation_code else ""
        date_start = _norm_date(query.issue_date_start)
        date_end   = _norm_date(query.issue_date_end)
        max_rec    = query.max_records if query.max_records > 0 else 10_000_000

        results: list[pb2.ViolationRecord] = []

        with open(data_path, encoding="utf-8", errors="replace") as fh:
            header = fh.readline()  # skip header
            for raw_line in fh:
                if len(results) >= max_rec:
                    break
                cols = _split_csv(raw_line.rstrip("\n"))
                if len(cols) < 22:
                    continue

                county = cols[21].strip()
                if county not in county_set:
                    continue
                if vc_filter and cols[5].strip() != vc_filter:
                    continue

                raw_date = cols[4].strip()
                norm_date = _norm_date(raw_date)
                if date_start and norm_date < date_start:
                    continue
                if date_end and norm_date > date_end:
                    continue

                rec = pb2.ViolationRecord()
                rec.summons_number            = _safe_int64(cols[0])
                rec.plate_id                  = cols[1].strip()
                rec.registration_state        = cols[2].strip()
                rec.plate_type                = cols[3].strip()
                rec.issue_date                = raw_date
                rec.violation_code            = _safe_int32(cols[5])
                rec.vehicle_body_type         = cols[6].strip()
                rec.vehicle_make              = cols[7].strip()
                rec.issuing_agency            = cols[8].strip()
                rec.street_code1              = _safe_int32(cols[9])
                rec.street_code2              = _safe_int32(cols[10])
                rec.street_code3              = _safe_int32(cols[11])
                rec.vehicle_expiration_date   = _safe_int32(cols[12])
                rec.violation_location        = cols[13].strip()
                rec.violation_precinct        = _safe_int32(cols[14])
                rec.issuer_precinct           = _safe_int32(cols[15])
                rec.issuer_code               = _safe_int64(cols[16])
                rec.issuer_command            = cols[17].strip()
                rec.issuer_squad              = cols[18].strip()
                rec.violation_time            = cols[19].strip()
                rec.time_first_observed       = cols[20].strip()
                rec.violation_county          = county
                rec.violation_front_opposite  = cols[22].strip() if len(cols) > 22 else ""
                rec.house_number              = cols[23].strip() if len(cols) > 23 else ""
                rec.street_name               = cols[24].strip() if len(cols) > 24 else ""
                rec.intersecting_street       = cols[25].strip() if len(cols) > 25 else ""
                rec.date_first_observed       = _safe_int32(cols[26]) if len(cols) > 26 else 0
                rec.law_section               = _safe_int32(cols[27]) if len(cols) > 27 else 0
                rec.sub_division              = cols[28].strip() if len(cols) > 28 else ""
                rec.violation_legal_code      = cols[29].strip() if len(cols) > 29 else ""
                rec.days_parking_in_effect    = cols[30].strip() if len(cols) > 30 else ""
                rec.from_hours_in_effect      = cols[31].strip() if len(cols) > 31 else ""
                rec.to_hours_in_effect        = cols[32].strip() if len(cols) > 32 else ""
                rec.vehicle_color             = cols[33].strip() if len(cols) > 33 else ""
                rec.unregistered_vehicle      = _safe_bool(cols[34]) if len(cols) > 34 else False
                rec.vehicle_year              = _safe_int32(cols[35]) if len(cols) > 35 else 0
                rec.meter_number              = cols[36].strip() if len(cols) > 36 else ""
                rec.feet_from_curb            = _safe_int32(cols[37]) if len(cols) > 37 else 0
                rec.violation_post_code       = cols[38].strip() if len(cols) > 38 else ""
                rec.violation_description     = cols[39].strip() if len(cols) > 39 else ""
                rec.no_standing_violation     = cols[40].strip() if len(cols) > 40 else ""
                rec.hydrant_violation         = cols[41].strip() if len(cols) > 41 else ""
                rec.double_parking_violation  = cols[42].strip() if len(cols) > 42 else ""
                results.append(rec)

        return results


# ---------------------------------------------------------------------------
# CSV / type helpers
# ---------------------------------------------------------------------------
def _split_csv(line: str) -> list[str]:
    cols, cur, in_q = [], [], False
    for ch in line:
        if ch == '"':
            in_q = not in_q
        elif ch == ',' and not in_q:
            cols.append("".join(cur))
            cur = []
        else:
            cur.append(ch)
    cols.append("".join(cur))
    return cols


def _norm_date(d: str) -> str:
    """MM/DD/YYYY → YYYY/MM/DD, or '' if unparseable."""
    d = d.strip()
    if not d:
        return ""
    parts = d.split("/")
    if len(parts) == 3:
        return f"{parts[2]}/{parts[0]}/{parts[1]}"
    return d


def _safe_int64(s: str) -> int:
    try:
        return int(s.strip())
    except (ValueError, AttributeError):
        return 0


def _safe_int32(s: str) -> int:
    try:
        return int(s.strip())
    except (ValueError, AttributeError):
        return 0


def _safe_bool(s: str) -> bool:
    return s.strip().lower() in ("1", "true", "yes", "y")


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------
def run(config_file: str) -> None:
    global _log_fh

    config = load_config(config_file)
    pid  = config["process_id"]
    host = config["listen_host"]
    port = config["listen_port"]

    os.makedirs("logs", exist_ok=True)
    _log_fh = open(f"logs/{pid}_worker.log", "a")

    servicer = WorkerServiceImpl(config)

    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    pb2_grpc.add_ParkingViolationServiceServicer_to_server(servicer, server)
    address = f"{host}:{port}"
    server.add_insecure_port(address)
    server.start()

    logger.info(f"\n*** Worker {pid} listening on {address} ***\n")
    _log("STARTED", extra=f"listening on {address}")

    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        logger.info(f"[Worker {pid}] shutting down")
        _log("SHUTDOWN", extra="process stopping")
        server.stop(grace=5)

    if _log_fh:
        _log_fh.close()


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <config_file>", file=sys.stderr)
        sys.exit(1)
    run(sys.argv[1])

File: src/servers/worker/worker_server.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <thread>
#include <mutex>
#include <chrono>

#include <grpcpp/grpcpp.h>
#include <grpcpp/server_builder.h>

#include "parking_violation_query.grpc.pb.h"
#include "../../common/config.hpp"
#include "../../common/parking_violations_loader.hpp"
#include "../../common/metrics.hpp"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::ServerWriter;
using grpc::Status;

using parkingviolation::ParkingViolationService;
using parkingviolation::QueryRequest;
using parkingviolation::QueryResponse;
using parkingviolation::DelegationRequest;
using parkingviolation::DelegationResponse;
using parkingviolation::HealthRequest;
using parkingviolation::HealthResponse;
using parkingviolation::CancelRequest;
using parkingviolation::CancelResponse;
using parkingviolation::ViolationRecord;

// Convert internal record to protobuf message
static void toProto(const ParkingViolationRecord& src, ViolationRecord* dst) {
    dst->set_summons_number(src.summons_number);
    dst->set_plate_id(src.plate_id);
    dst->set_registration_state(src.registration_state);
    dst->set_plate_type(src.plate_type);
    dst->set_issue_date(src.issue_date);
    dst->set_violation_code(src.violation_code);
    dst->set_vehicle_body_type(src.vehicle_body_type);
    dst->set_vehicle_make(src.vehicle_make);
    dst->set_issuing_agency(src.issuing_agency);
    dst->set_street_code1(src.street_code1);
    dst->set_street_code2(src.street_code2);
    dst->set_street_code3(src.street_code3);
    dst->set_vehicle_expiration_date(src.vehicle_expiration_date);
    dst->set_violation_location(src.violation_location);
    dst->set_violation_precinct(src.violation_precinct);
    dst->set_issuer_precinct(src.issuer_precinct);
    dst->set_issuer_code(src.issuer_code);
    dst->set_issuer_command(src.issuer_command);
    dst->set_issuer_squad(src.issuer_squad);
    dst->set_violation_time(src.violation_time);
    dst->set_time_first_observed(src.time_first_observed);
    dst->set_violation_county(src.violation_county);
    dst->set_violation_front_opposite(src.violation_front_opposite);
    dst->set_house_number(src.house_number);
    dst->set_street_name(src.street_name);
    dst->set_intersecting_street(src.intersecting_street);
    dst->set_date_first_observed(src.date_first_observed);
    dst->set_law_section(src.law_section);
    dst->set_sub_division(src.sub_division);
    dst->set_violation_legal_code(src.violation_legal_code);
    dst->set_days_parking_in_effect(src.days_parking_in_effect);
    dst->set_from_hours_in_effect(src.from_hours_in_effect);
    dst->set_to_hours_in_effect(src.to_hours_in_effect);
    dst->set_vehicle_color(src.vehicle_color);
    dst->set_unregistered_vehicle(src.unregistered_vehicle);
    dst->set_vehicle_year(src.vehicle_year);
    dst->set_meter_number(src.meter_number);
    dst->set_feet_from_curb(src.feet_from_curb);
    dst->set_violation_post_code(src.violation_post_code);
    dst->set_violation_description(src.violation_description);
    dst->set_no_standing_violation(src.no_standing_violation);
    dst->set_hydrant_violation(src.hydrant_violation);
    dst->set_double_parking_violation(src.double_parking_violation);
}

class WorkerServiceImpl final : public ParkingViolationService::Service {
public:
    WorkerServiceImpl(const ProcessConfig& cfg)
        : config_(cfg), loader_(cfg.data_path) {

        std::cout << "[Worker " << config_.process_id << "] starting"
                  << " (team=" << config_.team << ")\n";
        std::cout << "[Worker " << config_.process_id << "] listening on "
                  << config_.listen_host << ":" << config_.listen_port << "\n";
        std::cout << "[Worker " << config_.process_id << "] owned counties: ";
        for (const auto& c : config_.data_partitioning.owned_counties)
            std::cout << c << " ";
        std::cout << "\n";

        metrics::init(config_.process_id, config_.role);
    }

    // -----------------------------------------------------------------------
    // DelegateQuery — main entry point for workers
    // -----------------------------------------------------------------------
    Status DelegateQuery(ServerContext* context,
                         const DelegationRequest* request,
                         ServerWriter<DelegationResponse>* writer) override {

        std::cout << "\n[Worker " << config_.process_id
                  << "] delegation " << request->request_id()
                  << " from " << request->delegating_process() << "\n";

        metrics::log_event("RECV_DELEGATION", request->request_id(),
                           ++pending_, 1, -1, -1, request->delegating_process());

        // Deserialize original query
        QueryRequest query;
        if (!query.ParseFromString(request->original_query())) {
            --pending_;
            return Status(grpc::StatusCode::INVALID_ARGUMENT, "Failed to parse query");
        }

        // Determine counties this worker should process
        std::vector<std::string> counties = selectCounties(query);
        if (counties.empty()) {
            std::cout << "[Worker " << config_.process_id
                      << "] no matching counties — returning empty\n";
            --pending_;
            return Status::OK;
        }

        // Load matching records
        auto t0 = std::chrono::high_resolution_clock::now();
        auto records = loader_.loadData(
            counties,
            query.violation_code(),
            query.issue_date_start(),
            query.issue_date_end(),
            query.max_records()
        );
        auto load_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
                           std::chrono::high_resolution_clock::now() - t0).count();

        std::cout << "[Worker " << config_.process_id << "] loaded "
                  << records.size() << " records in " << load_ms << "ms\n";

        metrics::log_event("LOADED", request->request_id(),
                           pending_, 1, -1, (int)records.size(), "");

        // Stream in chunks
        int chunk_size = config_.chunk_config.default_chunk_size;
        int chunk_num  = 0;

        for (size_t i = 0; i < records.size(); i += chunk_size) {
            if (context->IsCancelled()) {
                metrics::log_event("CANCELLED", request->request_id(),
                                   pending_, 1, chunk_num, -1, "");
                --pending_;
                return Status::CANCELLED;
            }

            DelegationResponse resp;
            resp.set_request_id(request->request_id());
            resp.set_chunk_number(chunk_num);
            resp.set_is_final(false);
            resp.set_responding_process(config_.process_id);

            size_t end = std::min(i + (size_t)chunk_size, records.size());
            for (size_t j = i; j < end; ++j)
                toProto(records[j], resp.add_records());

            if (!writer->Write(resp)) {
                metrics::log_event("WRITE_FAIL", request->request_id(),
                                   pending_, 1, chunk_num, (int)resp.records_size(), "");
                --pending_;
                return Status::CANCELLED;
            }

            metrics::log_event("CHUNK_SENT", request->request_id(),
                               pending_, 1, chunk_num, (int)resp.records_size(),
                               config_.process_id);

            std::cout << "[Worker " << config_.process_id << "] sent chunk "
                      << chunk_num << " (" << resp.records_size() << " records)\n";
            ++chunk_num;
        }

        metrics::log_event("DONE", request->request_id(),
                           pending_, 1, chunk_num, (int)records.size(), "");
        --pending_;
        ++completed_;
        return Status::OK;
    }

    Status HealthCheck(ServerContext*,
                       const HealthRequest*,
                       HealthResponse* resp) override {
        resp->set_responding_process(config_.process_id);
        resp->set_is_healthy(true);
        resp->set_pending_requests(pending_);
        resp->set_active_workers(1);
        return Status::OK;
    }

    Status QueryViolations(ServerContext*,
                           const QueryRequest*,
                           ServerWriter<QueryResponse>*) override {
        return Status(grpc::StatusCode::UNIMPLEMENTED,
                      "Workers do not accept direct queries");
    }

    Status CancelQuery(ServerContext*,
                       const CancelRequest* req,
                       CancelResponse* resp) override {
        resp->set_request_id(req->request_id());
        resp->set_cancelled(true);
        resp->set_message("cancel acknowledged");
        return Status::OK;
    }

private:
    ProcessConfig          config_;
    ParkingViolationsLoader loader_;
    int pending_   = 0;
    int completed_ = 0;

    // Returns the subset of owned_counties that match the query's county filter.
    // If the query has no county filter, returns all owned_counties.
    std::vector<std::string> selectCounties(const QueryRequest& query) {
        const auto& owned = config_.data_partitioning.owned_counties;
        if (owned.empty()) return {};  // this process owns no data

        const std::string& qc = query.violation_county();
        if (qc.empty()) return owned;  // no filter — serve all owned

        // Check if any owned county matches the requested county
        for (const auto& c : owned) {
            if (c == qc) return {qc};
        }
        return {};  // query county not in this process's partition
    }
};

void RunWorker(const std::string& config_file) {
    ProcessConfig config = ConfigParser::loadConfig(config_file);
    WorkerServiceImpl service(config);

    std::string address = config.listen_host + ":" + std::to_string(config.listen_port);
    ServerBuilder builder;
    builder.AddListeningPort(address, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);

    auto server = builder.BuildAndStart();
    std::cout << "\n*** Worker " << config.process_id
              << " listening on " << address << " ***\n\n";
    server->Wait();
}

int main(int argc, char** argv) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <config_file>\n";
        return 1;
    }
    try {
        RunWorker(argv[1]);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        metrics::shutdown();
        return 1;
    }
    metrics::shutdown();
    return 0;
}